src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,216 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Razor;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    /// <summary>
+    /// Caches tokens on a per-line basis for documents based on semantic version.
+    /// </summary>
+    /// <remarks>
+    /// The cache makes the assumption that the tokens passed in are complete for the given range.
+    /// </remarks>
+    internal sealed class SemanticTokensCache
+    {
+        private const int MaxSemanticVersionPerDoc = 4;
+
+        // Nested cache mapping (URI -> (semanticVersion -> (line #s -> tokens on line)))
+        private readonly MemoryCache<DocumentUri, MemoryCache<VersionStamp, Dictionary<int, IReadOnlyList<int>>>> _cache = new();
```

Comment:
instead of an `IReadOnlyList<int>` it feels like ImmutableArray will be faster/cheaper.

Reply:
We don't know how long the tokens list will be until we process the tokens, so in order to do this we'd have to use a List + afterwards convert to ImmutableArray, which seems to be more work overall?

Reply:
you can use an ArrayBuilder + ToImmutable, which is our general IDE pattern :)

Reply:
Razor doesn't have ArrayBuilder atm, although I think that would be something to look into as a follow-up item. Maybe Roslyn could expose the type and Razor can consume it?